### PR TITLE
tests: add option to dump genesis files in E2E tests

### PR DESCRIPTION
**Pull request by bznein**
## Description

Adds new fields in the e2e config to enable dumping of Genesis files.

closes: #6032

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [x] Targeted PR against the correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#pull-request-targeting)).
- [x] Linked to GitHub issue with discussion and accepted design, OR link to spec that describes this work.
- [x] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/main/docs/build/building-modules/11-structure.md) and [Go style guide](../docs/dev/go-style-guide.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/main/testing/README.md#ibc-testing-package).
- [ ] Updated relevant documentation (`docs/`).
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [x] Provide a [conventional commit message](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#commit-messages) to follow the repository standards.
- [ ] Include a descriptive changelog entry when appropriate. This may be left to the discretion of the PR reviewers. (e.g. chores should be omitted from changelog)
- [x] Re-reviewed `Files changed` in the GitHub PR explorer.
- [ ] Review `SonarCloud Report` in the comment section below once CI passes.

## Summary by CodeRabbit

- **Tests**
	- Improved test suites in multiple modules by adding Genesis debug export configuration for enhanced debugging and diagnostics.
- **Chores**
	- Renamed function `getE2EDir` to `GetE2EDir` in `diagnostics.go` for consistent naming.
	- Introduced new constants, variables, methods, and structs for Genesis debug management in `testconfig.go`.
- **Chores**
	- Updated tool naming conventions in `diagnostics.go` for better clarity.

### CODE REVIEW

File: e2e/scripts/run-e2e.sh [hunk 1]
```diff
@@ -59,7 +59,7 @@ _verify_dependencies
 
 # if the dev configs directory is present, enable fzf completion to select a test config file to use.
 if [[ -d "dev-configs"  ]]; then
-  export E2E_CONFIG_PATH="$(pwd)/dev-configs/$(_select_test_config)"
+  export E2E_CONFIG_PATH="/workspaces/ibc-go/e2e/dev-configs/config.yaml"
```
Comment (bznein) [inline e2e/scripts/run-e2e.sh:62]: This needs to be reverted before review. I had to do it as I don't have fzf available on github codespaces

File: e2e/testsuite/testsuite.go [hunk 2]
```diff
@@ -67,6 +70,47 @@ func newPath(chainA, chainB ibc.Chain) pathPair {
...
+	}
+
+	// Set the export path.
+	exportPath := cfg.ExportFilePath
+
+	// If no path is provided, use the default (e2e/diagnostics/genesis.json).
+	if exportPath == "" {
+		e2eDir, err := diagnostics.GetE2EDir(t)
+		if err != nil {
+			s.Fail("can't get e2edir: %w", err)
```
[the remark this answers is not in the document]
Comment (bznein) [reply e2e/testsuite/testsuite.go:89]: Thank you kind rabbit!

File: e2e/testsuite/testsuite.go [hunk 3]
```diff
@@ -67,6 +69,48 @@ func newPath(chainA, chainB ibc.Chain) pathPair {
...
+		exportPath = gopath.Join(e2eDir, defaultGenesisExportPath)
+	}
+
+	if !gopath.IsAbs(exportPath) {
+		wd, err := os.Getwd()
+		s.Require().NoError(err, "can't get working directory")
+		exportPath = gopath.Join(wd, exportPath)
+	}
+
+	t.Setenv("EXPORT_GENESIS_FILE_PATH", exportPath)
```
Comment (chatton) [inline e2e/testsuite/testsuite.go:101]: maybe add a link to interchaintest code for where this is used

File: e2e/testsuite/testconfig.go [hunk 4]
```diff
@@ -68,8 +68,14 @@ const (
 	// defaultConfigFileName is the default filename for the config file that can be used to configure
 	// e2e tests. See sample.config.yaml as an example for what this should look like.
 	defaultConfigFileName = ".ibc-go-e2e-config.yaml"
+
+	// defaultGenesisExportPath is the default path to which Genesis debug files will be exported to.
+	defaultGenesisExportPath = "diagnostics/genesis.json"
```
Comment (chatton) [inline e2e/testsuite/testconfig.go:73]: if we end up moving the GetE2EPath to internal, maybe we can move this constant there too.

File: e2e/testsuite/testsuite.go [hunk 5]
```diff
@@ -67,6 +69,48 @@ func newPath(chainA, chainB ibc.Chain) pathPair {
...
+	if !cfg.DumpGenesisDebugInfo {
+		return
+	}
+
+	// Set the export path.
+	exportPath := cfg.ExportFilePath
+
+	// If no path is provided, use the default (e2e/diagnostics/genesis.json).
+	if exportPath == "" {
+		e2eDir, err := diagnostics.GetE2EDir(s.T())
```
[the remark this answers is not in the document]
Comment (bznein) [reply e2e/testsuite/testsuite.go:90]: Good point, done!

File: e2e/internal/directories/directories.go [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+	count := 0
+	for ; !strings.HasSuffix(wd, e2eDir) || count > maxAttempts; wd = path.Dir(wd) {
+		count++
+	}
+
+	// arbitrary value to avoid getting stuck in an infinite loop if this is called
+	// in a context where the e2e directory does not exist.
+	if count > maxAttempts {
+		return "", fmt.Errorf("unable to find e2e directory after %d tries", maxAttempts)
+	}
```
[the remark this answers is not in the document]
Comment (chatton) [reply e2e/internal/directories/directories.go:37]: it is not robust but we don't need it to work in other environments!

File: e2e/testsuite/testsuite.go [hunk 7]
```diff
@@ -4,6 +4,8 @@ import (
 	"context"
 	"errors"
 	"fmt"
+	"os"
+	gopath "path"
```
[the remark this answers is not in the document]
Comment (bznein) [reply e2e/testsuite/testsuite.go:8]: Good catch and yes! The reason is that at line 175 we have a function that uses a variable named path, and this made the linter sad :(

Of course, now that I think about it, I could have easily just changed the variable instead, which is probably cleaner. WDYT?

[the remark this answers is not in the document]
Comment (bznein) [reply e2e/testsuite/testsuite.go]: Done!

File: e2e/testsuite/testsuite.go [hunk 8]
```diff
@@ -67,6 +70,49 @@ func newPath(chainA, chainB ibc.Chain) pathPair {
 	}
 }
 
+func (s *E2ETestSuite) SetupTest() {
+	s.ConfigureGenesisDebugExport()
```
[the remark this answers is not in the document]
Comment (chatton) [reply e2e/testsuite/testsuite.go:74]: yeah, basically just a way to make sure that env vars get set for each specific test based on the config.

File: e2e/testsuite/testsuite.go [hunk 9]
```diff
@@ -67,6 +70,49 @@ func newPath(chainA, chainB ibc.Chain) pathPair {
...
+}
+
+// ConfigureGenesisDebugExport sets, if needed, env variables to enable exporting of Genesis debug files.
+func (s *E2ETestSuite) ConfigureGenesisDebugExport() {
```
[the remark this answers is not in the document]
Comment (bznein) [reply e2e/testsuite/testsuite.go]: Done!

**Pull request merged**